### ISSUE TITLE
BatchNorm{1-2-3}d are redundant

### ISSUE THREAD

**Issue by fmassa** [pytorch/pytorch#2628]
Since https://github.com/pytorch/pytorch/pull/2020, `nn.BatchNorm1d`/`nn.BatchNorm2d`/`nn.BatchNorm3d` are completely redundant, as there are no more size checks that are performed in the input tensor
I see two options here:
* add the size-checks back for each class;
* deprecate BatchNorm{1-2-3}d and use instead BatchNorm.

We already have only a single functional interface for BatchNorm, so option 2 might be worth considering.

cc. @alykhantejani who removed the size checks

**Comment by alykhantejani** [comment]
I'm not sure the size checks are necessary as we have a single functional interface for batch_norm and these checks aren't performed there (as we don't know if it's 1D, 2D or 3D).  I think we should remove the dead code in BatchNorm{1-2-3}d that overrides `_check_input_dim` which now never gets called (sorry I missed this in my earlier PR).

I'm inclined towards keeping the modules and functional interface as aligned as possible, so maybe just having a single BatchNorm module would be nice which is similar to how it's done in TF and Keras.

After discussing it a bit with @fmassa perhaps we can keep the classes for backwards compatibility and add deprecation warning inside BatchNorm{1-2-3}d saying the user should use `BatchNorm` instead, we could/should also remove the docs for BatchNorm{1-2-3}d and just have docs for `BatchNorm` saying it accepts different input shapes. 

Happy to send a PR for these changes, WDYT?

**Comment by soumith** [comment]
isn't the 1D case ambiguous between the Spatial version and the Local version? i.e. one for Conv1d outputs and one for using with output of Linear?

**Comment by fmassa** [comment]
I think it should be fine because the output of `Linear` is 2d, while `Conv1d` is 3d, so those cases are currently handled in both [`cudnn`](https://github.com/pytorch/pytorch/blob/master/torch/csrc/cudnn/BatchNorm.cpp#L68-L75) as well as in THNN (because it always take the 2nd dimension for the channel normalization) I think? The fix from https://github.com/pytorch/pytorch/pull/423 to handle `Conv1d` didn't add anything special to the backends, just an assertion on the dimension.
